espnet/nets/pytorch_backend/k2_ctc.py

```diff
@@ -0,0 +1,211 @@
+#!/usr/bin/env python3
+# Copyright (c)  2021  Mobvoi Inc. (authors: Yaguang Hu)
+# Apache 2.0  (http://www.apache.org/licenses/LICENSE-2.0)
+
+"""k2 based ctc loss implementation."""
+
+from functools import lru_cache
+import math
+from typing import List
+from typing import Tuple
+
+import k2
+import logging
+import torch
+
+
+class K2CTCLoss(torch.nn.Module):
+    """K2CTCLoss module."""
+
+    def __init__(
+        self,
+        odim: int,
+        reduction: str = "sum",
+        device: torch.device = torch.device("cpu"),
+    ) -> None:
+        """K2CTCLoss module.
+
+        Args:
+          odim:
+            Number of classes (including blank).
+          reduction:
+            Specifies the reduction to apply to the output. Default: 'sum'
+          device:
+            An instance of `torch.device`. Default: 'cpu'
+        """
+        torch.nn.Module.__init__(self)
+        self.device = device
+        self.reduction = reduction
+        if reduction not in ["sum", "none"]:
+            logging.error(
+                f"k2 ctc loss reduction type: {reduction} is not supported yet,"
+                "change it to be 'sum'"
+            )
+            self.reduction = "sum"
+        self.graph_compiler = CtcTrainingGraphCompiler(odim, device=device)
+
+    def forward(
+        self,
+        log_probs: torch.Tensor,
+        targets: torch.Tensor,
+        input_lengths: torch.Tensor,
+        target_lengths: torch.Tensor,
+    ) -> torch.Tensor:
+        """k2 ctc forward function.
+
+        Args:
+          log_probs:
+            Tensor of size (T, N, C), where T is input length, N is batch size,
+            and C is number of classes (including blank). Note that it must be
+            log probabilities.
+          targets:
+            Tensor of size (sum(target_lengths)). It represent corresponding labels
+            of log_probs.
+          input_lengths:
+            Tuple or tensor of size (N). It represent the lengths of the inputs.
+          target_lengths:
+            Tuple or tensor of size (N). It represent lengths of the targets.
+        """
+        log_probs = log_probs.permute(
+            1, 0, 2
+        )  # now log_probs is [N, T, C]  batchSize x seqLength x alphabet_size
+        supervision_segments = torch.stack(
+            (
+                torch.tensor(range(input_lengths.shape[0])),
+                torch.zeros(input_lengths.shape[0]),
+                input_lengths,
```

Comment:
Are elements of `input_lengths` sorted in descending order?

If not, I would recommend you to sort them. Please refer to
https://github.com/k2-fsa/snowfall/blob/bbcf1ec1185179b9b502287c51b46fbc2accb8d0/egs/librispeech/asr/simple_v1/ctc_train.py#L80-L81

```
    indices = torch.argsort(supervision_segments[:, 2], descending=True)
    supervision_segments = supervision_segments[indices]
```

Reply:
If this constraint turns out to be a hassle, we could perhaps introduce into the C++ a recursion, that checks whether they are sorted by length, and if not, sorts them by length, does the intersection and then indexes them to get them back in the correct order.